### PR TITLE
Ryna-LA100.socioeconomics_china

### PR DESCRIPTION
**Pull request by rynacui**
First chunk of GCAM-China, try to find out if everything works as we rebase to v1.0

### CODE REVIEW

File: inst/extdata/gcam-china/gdp_mil10usd_province.csv [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# File: gdp_mil10usd_province.csv
+# Title: GDP by province
+# Units: million 2010 USD
+# Description: Provincial GDP by historical and future years
+# Source: Citation or source if available
```
Comment (rynacui) [inline inst/extdata/gcam-china/gdp_mil10usd_province.csv]: data source needed @sypnl

File: inst/extdata/gcam-china/population_thous_province.csv [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# File: population_thous_province.csv
+# Title: Population by province
+# Units: 1000 persons
+# Description: Provincial population by historical and future years
+# Source: Citation or source if available
```
Comment (rynacui) [inline inst/extdata/gcam-china/population_thous_province.csv]: data source needed @sypnl

Comment (sypnl) [reply inst/extdata/gcam-china/population_thous_province.csv]: Historical data (1975-2013) on national population were obtained from the 2014 China Statistics Yearbook and the China Compendium of 60-year Statistics (1949-2008). UN medium variant growth rates were applied to project future population in China from 2014 to 2100.

File: inst/extdata/gcam-china/gdp_mil10usd_province.csv (hunk 1 above)
Comment (sypnl) [reply inst/extdata/gcam-china/gdp_mil10usd_province.csv]: @rynacui Historical data (1975-2013) on GDP per capita were obtained from the 2014 China Statistics Yearbook and the China Compendium of 60-year Statistics (1949-2008) as in current CNY. The numbers were then all converted to 2010CNY using the World Bank GDP deflators for China (World Development Indicators Database, 2014). All the numbers were converted to 2010USD using the annual average USD-CNY exchange rate in 2010. IMF projected growth rates were applied to project GDP per capita from 2014 to 2020 assuming the same growth rate in 2019 for 2020.  Growth rates from 2021 to 2100 came from the OECD SSP2 Database (IIASA SSP Database, 2015).

Comment (rynacui) [reply inst/extdata/gcam-china/gdp_mil10usd_province.csv]: This input file is total GDP. Are total GDP calculated from pcap GDP and population?

Comment (sypnl) [reply inst/extdata/gcam-china/gdp_mil10usd_province.csv]: yep.

File: R/module-helpers.R [hunk 3]
```diff
@@ -625,3 +625,52 @@ downscale_FAO_country <- function(data, country_name, dissolution_year, years =
...
+#' @param replace Replace old province name to new name columns? Logical
+#' @return Data with new province name columns added or replaced.
+map_province_name <- function(data, mapping, map_name, replace = FALSE) {
+
+  orig.names <- old.name <- old.index <- new.names <- NULL  # silence package check.
+
+  # Sanity checks
+  assert_that(tibble::is_tibble(data))
+  assert_that(tibble::is_tibble(mapping))
+  assert_that(is.character(map_name))
```
Comment (bpbond) [inline R/module-helpers.R]: 👍 nice defensive programming @rynacui

File: R/module-helpers.R [hunk 4]
```diff
@@ -625,3 +625,52 @@ downscale_FAO_country <- function(data, country_name, dissolution_year, years =
...
+
+  orig.names <- names(data)
+  # Some error checking to make sure the mapping names match up.
+  error <- ""
+  old.name <- names(data)[names(data) %in% names(mapping)]
+  if(length (old.name) != 1) {
+    error <- paste( error, "Could not map province: data has zero or many matching column names in mapping.", sep="\n")
```
Comment (bpbond) [inline R/module-helpers.R]: No space after parenthesis pls

Comment (bpbond) [inline R/module-helpers.R]: Why not just `stop(error, "\nCould not map province: data has zero or many matching column names in mapping.")`  ?

File: R/module-helpers.R [hunk 5]
```diff
@@ -625,3 +625,52 @@ downscale_FAO_country <- function(data, country_name, dissolution_year, years =
...
+  }
+  if(!any(map_name %in% names(mapping))) {
+    error <- paste(error, "Could not map province: one or more map.names do not match columnnames in mapping.", sep="\n")
+  }
+  if(error != "") {
+    stop(error)
```
Comment (bpbond) [inline R/module-helpers.R]: Yeah it seems like just `stop` immediately in either of the cases above, so don't need `error` at all.

File: R/zchunk_LA100.Socioeconomics_china.R [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    population_thous_province %>%
+      gather(year, pop, matches("[0-9]{4}")) %>%
+      mutate(year = as.integer(year)) %>%
+      # Complete data with all provinces and years and fill missing values
+      complete(nesting(province.name), year = socio_hist_years) %>%
+      arrange(province.name, year) %>%
+      group_by(province.name) %>%
+      mutate(pop = approx_fun(year, pop, rule = 2)) %>%
+      ungroup %>%
+      # Replace full provice names with abbreviation
```
Comment (bpbond) [inline R/zchunk_LA100.Socioeconomics_china.R]: "province"

File: R/module-helpers.R (hunk 5 above)
Comment (rynacui) [reply R/module-helpers.R]: Got it. Thank you!

File: inst/extdata/gcam-china/province_names_mappings.csv [hunk 7]
```diff
@@ -0,0 +1,38 @@
+# File: province_names_mappings.csv
+# Title: Province alternate names mapping file
+# Units: NA
+# Description: Map province alternate names, climate region and grid region
```
Comment (sypnl) [inline inst/extdata/gcam-china/province_names_mappings.csv]: # Description: Map province alternate names, administrative region, climate region, and grid region

**Comment by bpbond** [comment]
>Could you add an additional comment?

Please capture this by [creating a new issue](https://github.com/JGCRI/gcamdata/issues/new) (feel free to create a new `gcam-china` tag too).

**Comment by rynacui** [comment]
@bpbond looks good to you?

File: R/module-helpers.R [hunk 8]
```diff
@@ -625,3 +625,48 @@ downscale_FAO_country <- function(data, country_name, dissolution_year, years =
...
+
+  # Do the name mapping
+  data <- left_join_error_no_match(data, mapping %>% select(old.name, map_names), by = old.name)
+
+  # Attempt to maintain the orgininal column ordering, if we
+  # are replacing then the new column use that position otherwise
+  # will just be added to the end.
+  if(replace) {
+    old.index <- match(old.name, orig.names)
+    new.names <- c(orig.names[1:length(orig.names) < old.index], map_names, orig.names[1:length(orig.names) > old.index])
```
Comment (bpbond) [inline R/module-helpers.R]: Instead of `1:length(orig.names)` it's safer to use `seq_along(orig.names)`

**Comment by rynacui** [comment]
I don't think I can merge, and THANKS! @bpbond

**Pull request merged**